server/pairing/client.go

```diff
@@ -38,31 +39,53 @@ type BaseClient struct {
 	challengeTaker *ChallengeTaker
 }
 
-func findServerCert(c *ConnectionParams) (*url.URL, *x509.Certificate, error) {
-	netIps, err := server.FindReachableAddressesForPairingClient(c.netIPs)
-	if err != nil {
-		return nil, nil, err
-	}
+func findServerCert(c *ConnectionParams, reachableIPs []net.IP) (*url.URL, *x509.Certificate, error) {
 	var baseAddress *url.URL
 	var serverCert *x509.Certificate
-	var certErrs error
-	for _, ip := range netIps {
-		u := c.BuildURL(ip)
-
-		serverCert, err = getServerCert(u)
-		if err != nil {
-			var certErr string
-			if certErrs != nil {
-				certErr = certErrs.Error()
+	var once sync.Once
+	errCh := make(chan struct {
+		ip  net.IP
+		err error
+	}, len(reachableIPs))
+	certCh := make(chan *x509.Certificate, len(reachableIPs))
+	urlCh := make(chan *url.URL, len(reachableIPs))
+	done := make(chan bool)
+
+	for _, ip := range reachableIPs {
+		go func(ip net.IP) {
+			u := c.BuildURL(ip)
+			cert, err := getServerCert(u)
+			if err != nil {
+				errCh <- struct {
+					ip  net.IP
+					err error
+				}{ip: ip, err: fmt.Errorf("connecting to '%s' failed: %s", u, err.Error())}
+				return
+			}
+			// If no error, send the results to their respective channels
+			urlCh <- u
+			certCh <- cert
+			once.Do(func() { close(done) }) // signal success and close the done channel
+		}(ip)
+	}
+
+	// Keep track of error counts
+	errorCount := 0
+	var combinedErrors string
+	for {
+		select {
+		case <-done:
+			baseAddress = <-urlCh
+			serverCert = <-certCh
+			return baseAddress, serverCert, nil
+		case ipErr := <-errCh:
+			errorCount++
+			combinedErrors += fmt.Sprintf("IP %s: %s; ", ipErr.ip, ipErr.err)
```

Comment:
`%w` can be used for error directly

Reply:
I don't think we should use `%w` here since they're not `wrap` relation, WDYT? 🙂

Reply:
Unfortunately wrapping only applies to errors on different call depths. In `go 1.20` you can use `errors.Join` https://pkg.go.dev/errors@master#Join for errors generated on the same call depth.

At the moment `status-go` is using `go 1.19`, so we don't have access to `errors.Join` which would be a nice solution.

Reply:
The desktop app is using `go1.20.4`.